explorer/update.go

```diff
@@ -0,0 +1,189 @@
+package explorer
+
+import (
+	"sync"
+
+	"go.sia.tech/core/chain"
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/types"
+)
+
+// A Store is a database that stores information about elements, contracts,
+// and blocks.
+type Store interface {
+	AddSiacoinElement(sce types.SiacoinElement) error
+	AddSiafundElement(sce types.SiafundElement) error
+	AddFileContractElement(fce types.FileContractElement) error
+	RemoveElement(id types.ElementID) error
+	SiacoinElement(id types.ElementID) (types.SiacoinElement, error)
+	SiafundElement(id types.ElementID) (types.SiafundElement, error)
+	FileContractElement(id types.ElementID) (types.FileContractElement, error)
+	BlockFacts(height uint64) (BlockFacts, error)
+	AddBlockFacts(height uint64, facts BlockFacts) error
+	RemoveBlockFacts(height uint64) error
+	BlockHeight() (uint64, error)
+	SetBlockHeight(height uint64) error
+}
+
+// An Explorer contains a database storing information about blocks, outputs,
+// contracts.
+type Explorer struct {
+	db Store
+	mu sync.Mutex
+	vc consensus.ValidationContext
+}
+
+// ProcessChainApplyUpdate implements chain.Subscriber.
+func (e *Explorer) ProcessChainApplyUpdate(cau *chain.ApplyUpdate, _ bool) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	oldFacts, err := e.db.BlockFacts(e.vc.Index.Height)
```

Comment:
consider adding a `tipStats  ChainStats` field to `Explorer` instead of retrieving them from the db each time